### DISCUSSION TITLE
Can I use Devexpress  Reporting

### DISCUSSION THREAD

**Discussion by eissa1005** [serenity-is/Serenity#5795] [Help] [answered]
I want use Devexpress  Reporting  In designing reports instead of html and css

**Comment by VictorTomaili** [comment]
There is no restriction about using any of the tech, library, or code in the serenity platform. The actual question is how? Implementation is depended on which tech and version. We have some customers who use DevExpress mixed with the serenity platform without a problem.

**Comment by eissa1005** [comment]
@VictorTomaili 
I understand this so use some of the reports the way it is used in serenity platform but  How do I Create a Master-Detail Report with a Detail Report Band and How do I design a similar report so that it includes groups for a specific category and also how
 to do I design a report with the ability to change some properties such as changing the Header color of the Report, font, color, logo and the ability to Review the report and not print it directly or convert it to a file PDF

**Reply by VictorTomaili** [reply]
Check [wkhtmltopdf](https://wkhtmltopdf.org/) tool documentation and [jsPdf](http://raw.githack.com/MrRio/jsPDF/master/docs/) documentation what used by serenity. 
Also, check this [HTML feature](https://www.w3schools.com/tags/att_style_media.asp) for preview a report in HTML then print it.

**Reply by eissa1005** [reply]
I will search for it, but I have a lot of ready-made reports that are designed 
 by Devexpress Report  I do not want to re-design them. I have prepared the DevExpress library so that the report can now be viewed with Serenity

**Reply by eissa1005** [reply]
![Untitled](https://user-images.githubusercontent.com/20481378/121812404-f344c400-cc67-11eb-9c10-40747ea82d79.png)

**Reply by eissa1005** [reply]
@VictorTomaili 
But there is a problem at the end of the report, can you help me solve it?
problem  => dx-analytics-core.js:7146 Cannot create a document for the current report (400 Bad Request)

**Reply by VictorTomaili** [reply]
If you want custom implementation support please contact us via email, check [serenity.is](https://serenity.is)

**Reply by eissa1005** [reply]
on this email sales@serenity.is

**Reply by ra-bedo11** [reply]
ازاى حليت المشكله دى لان انا واقع فيها مع Serenity asp.net core 5 ?

**Comment by eissa1005** [comment, marked as answer]
@VictorTomaili 
I sent the email but this problem I solved
![Untitled](https://user-images.githubusercontent.com/20481378/121821324-a4635280-cc98-11eb-8ad8-0cc489138232.png)

**Reply by VictorTomaili** [reply]
If you solved it's great, I'm happy to hear that. 👍

**Reply by eissa1005** [reply]
@VictorTomaili 
Thank you very much
I writed  suggestion with mail  that I sent   I would be very happy if you would have can help  me for it if it was available for Serenity

**Reply by reach2rv** [reply]
@eissa1005 can you please share the steps you took to implement devxpress.. though I have extended serenity to generate reports in grid as output.  I am still curious  to know other options for reporting

Thanks in advance

**Reply by eissa1005** [reply]
@reach2rv 
https://docs.devexpress.com/XtraReports/400044/web-reporting/asp-net-core-reporting/document-viewer-in-asp-net-applications/quick-start/create-an-aspnet-core-application-with-a-document-viewer
